src/Parsers/IAST.h

```diff
@@ -238,6 +238,9 @@ class IAST : public std::enable_shared_from_this<IAST>, public TypePromotion<IAS
         }
 
         void writeIdentifier(const String & name) const;
+        // Quote identifier `name` even when `always_quote_identifiers` is false.
+        // If `identifier_quoting_style` is `IdentifierQuotingStyle::None`, quote it with `IdentifierQuotingStyle::Backticks`
+        void quoteIdentifier(const String & name) const;
```

Comment:
This function is used to replace these two following calls:
1. [format_settings.ostr << backQuote(name);](https://github.com/ClickHouse/ClickHouse/blob/master/src/Parsers/ASTColumnDeclaration.cpp#L70)
2. [settings.ostr << backQuote(name);](https://github.com/ClickHouse/ClickHouse/blob/master/src/Parsers/ASTDictionaryAttributeDeclaration.cpp#L38)

Both use `backQuote` to quote `name` regardless of the `FormatSettings`. Even when the style is `IdentifierQuotingStyle::None`, `name` is still quoted with `Backticks`.
There is a comment which explains the reasons in `ASTColumnDeclaration::formatImpl`

```
/// We have to always backquote column names to avoid ambiguity with INDEX and other declarations in CREATE query.
```

So I created this `quoteIdentifier` function to always quote the identifier for this case. The style will be the same as `output_format_identifier_quoting_style`. However, when `output_format_identifier_quoting_style` is `IdentifierQuotingStyle::None`, it will use `Backticks` as a fallback case.

Regarding the tests, because the default values of `identifier_quoting_style` and `always_quote_identifiers` in `FormatSettings` are `IdentifierQuotingStyle::Backticks` and `false` respectively, when calling `formatWithPossiblyHidingSensitiveData`, I passed the same values. So the result would be matched with the reference.

Reply:
Ah, yes, you're right, that's a nice catch, I missed that.

If we want to support a setting that completely turns off unnecessary quotes, we can add missing keywords to https://github.com/ClickHouse/ClickHouse/blob/224e1940ae575c0b217b7f733576ae4dbce0fa08/src/IO/WriteHelpers.cpp#L88-L101
to avoid ambiguity in `ASTColumnDeclaration`

But I'm not sure if we should do it because it looks fragile: it may stop working when a new ambiguity is added without updating `writeProbablyQuotedStringImpl`. On the other hand, we have an explicit list of keywords: https://github.com/ClickHouse/ClickHouse/blob/2a6f498b7751df0e37f281136d524f542f7910c2/src/Parsers/CommonParsers.h#L572

Reply:
I think columns are quoted intentionally to avoid ambiguity with INDEX when creating tables, and we should always quote them.
BTW, is the PR good to be merged?